ssl/quic/qlog.c

```diff
@@ -0,0 +1,702 @@
+        BIO_free_all(bio);
+        return 0;
+    }
+
+    return 1;
+}
+
+int ossl_qlog_flush(QLOG *qlog)
+{
+    if (qlog == NULL)
+        return 1;
+
+    return ossl_json_flush(&qlog->json);
+}
+
+int ossl_qlog_set_event_type_enabled(QLOG *qlog, uint32_t event_type,
+                                     int enabled)
+{
+    if (qlog == NULL || event_type >= QLOG_EVENT_TYPE_NUM)
+        return 0;
+
+    bit_set(qlog->enabled, event_type, enabled);
+    return 1;
+}
+
+int ossl_qlog_enabled(QLOG *qlog, uint32_t event_type)
+{
+    if (qlog == NULL)
+        return 0;
+
+    return bit_get(qlog->enabled, event_type);
+}
+
+/*
+ * Event Lifecycle
+ * ===============
+ */
+static void write_str_once(QLOG *qlog, const char *key, char **p)
+{
+    if (*p == NULL)
+        return;
+
+    ossl_json_key(&qlog->json, key);
+    ossl_json_str(&qlog->json, *p);
+
+    OPENSSL_free(*p);
+    *p = NULL;
+}
+
+static void qlog_event_seq_header(QLOG *qlog)
+{
+    if (qlog->header_done)
+        return;
+
+    ossl_json_object_begin(&qlog->json);
+    {
+        ossl_json_key(&qlog->json, "qlog_version");
+        ossl_json_str(&qlog->json, "0.3");
+
+        ossl_json_key(&qlog->json, "qlog_format");
+        ossl_json_str(&qlog->json, "JSON-SEQ");
+
+        write_str_once(qlog, "title", (char **)&qlog->info.title);
+        write_str_once(qlog, "description", (char **)&qlog->info.description);
+
+        ossl_json_key(&qlog->json, "trace");
+        ossl_json_object_begin(&qlog->json);
+        {
+            ossl_json_key(&qlog->json, "common_fields");
+            ossl_json_object_begin(&qlog->json);
+            {
+                ossl_json_key(&qlog->json, "time_format");
+                ossl_json_str(&qlog->json, "delta");
+
+                ossl_json_key(&qlog->json, "protocol_type");
+                ossl_json_array_begin(&qlog->json);
+                {
+                    ossl_json_str(&qlog->json, "QUIC");
+                }
+                ossl_json_array_end(&qlog->json);
+
+                write_str_once(qlog, "group_id", (char **)&qlog->info.group_id);
+
+                ossl_json_key(&qlog->json, "system_info");
+                ossl_json_object_begin(&qlog->json);
+                {
+                    ossl_json_key(&qlog->json, "process_id");
+                    ossl_json_u64(&qlog->json, (uint64_t)getpid());
+                }
+                ossl_json_object_end(&qlog->json);
+            }
+            ossl_json_object_end(&qlog->json);
+
+            ossl_json_key(&qlog->json, "vantage_point");
+            ossl_json_object_begin(&qlog->json);
+            {
+                ossl_json_key(&qlog->json, "type");
+                ossl_json_str(&qlog->json, qlog->info.is_server
+                                    ? "server" : "client");
+            }
+            ossl_json_object_end(&qlog->json);
+        }
+        ossl_json_object_end(&qlog->json);
+    }
+    ossl_json_object_end(&qlog->json);
+
```

Comment:
It seems moot since if an error occurs output is aborted anyway. So there would be no reason to believe trying to serialize the header again later would be of any use or lead to coherent output (e.g. the header may already have been partially output and then an I/O error occurs).